numpy/lib/index_tricks.py

```diff
@@ -589,6 +590,336 @@ def __iter__(self):
+            step = (step,)
+        elif not isinstance(step, tuple):
+            step = tuple(step)
+
+        if len(start) != len(stop):
+            raise ValueError(
+                "``start`` and ``stop`` must have the same length.")
+
+        if len(step) != len(stop):
+            raise ValueError(
+                "``start``, ``stop``, and ``step`` must have the same length.")
+
+        self._start = start
+        self._stop = stop
+        self._step = step
+        self._ranges = tuple(range(*r) for r in zip(start, stop, step))
+
+    @property
+    def start(self):
+        """
+        The value of the start parameter (or ``(0,) * len(stop)`` if the
+        parameter was not supplied)
+        """
+        return self._start
+
+    @property
+    def stop(self):
+        """
+        The provided ``stop`` parameter as a tuple.
+        """
+        return self._stop
+
+    @property
+    def step(self):
+        """
+        The value of the ``step`` parameter (or ``(1,) * len(stop)`` if the
+        parameter was not supplied)
+        """
+        return self._step
+
+    def count(self, x):
+        """
+        Return number of occurrences of a tuple.
+
+        raises ``ValueError`` if the value is not found.
+        """
+        return 1 if x in self else 0
+
+    def index(self, x, start=None, stop=None):
+        """
+        Return the index of the provided value.
+
+        raises ``ValueError`` if the index is not found.
+
+        Parameters
+        ----------
+        x: tuple of ints, or int
+            Value for which to look for in the ndrange object.
+
+        Returns
+        -------
+        index: tuple of ints
+            Index of the value x.
+
+        """
+        # Note: It was empiracally found that indexing range objects does not
+        # support the additional ``start`` and ``stop`` parameters even in
+        # Python 3.7.0.
+        # A relevant bug report is found https://bugs.python.org/issue34848
+        if not isinstance(x, tuple):
+            x = (x,)
+
+        if len(x) != len(self._ranges):
+            raise ValueError("{} is not in range".format(x))
+
+        return tuple(r.index(i) for r, i in zip(self._ranges, x))
+
+    @property
+    def shape(self):
+        return tuple(len(r) for r in self._ranges)
+
+    def __iter__(self):
+        return itertools.product(*(iter(r) for r in self._ranges))
+
+    def __reversed__(self):
+        return itertools.product(*(reversed(r) for r in self._ranges))
+
+    def __contains__(self, index):
+        # Should we accept lists/arrays in the containment operation?
+        if not isinstance(index, tuple):
+            index = (index,)
+
+        if len(index) != len(self._ranges):
+            return False
+
+        return all(i in r for i, r in zip(index, self._ranges))
+
+    def __getitem__(self, sl):
+        # TODO: what is the correct way to handle non-tuple inputs?
```

Comment:
The behavior here seems good. But maybe we could add a nicer error message? Currently if you pass in a list, you get:
```python
>>> np.ndrange((1,2,3))[[0,0,0]]
TypeError: range indices must be integers or slices, not list
```
It would be nice if that said something about how the index can be a tuple.

Related: #9686. We want to encourage people to index with tuples instead of lists, so some similar logic should apply here.

Reply:
I don't know how to get the `not list` part. Checking for types in python is hard....